File: R_query.py
""" 
    11-27-24: Python version of PHP based Query Builder
"""

class r_query() :

    def set_fields_and_tables(self, array) :
        ''' 
            Pass a dict of field => Table values of ALL fields for query, even if they
            are not in the SELECT criteria
        '''
        self.fields_and_tables = array
    
    def set_joins_tables(self, array):
        ''' 
            Pass a dict of tables that require specific joins via RMSFILENUM and record types
        '''
        self.joins = array
    
    def set_selected_fields(self, array):
        #List of fields for SELECT criteria
        self.selections = array 
    
    def create_table_index_selections_and_fields(self):
        # Creates a master index through a DICT of all tables and fields to be used and their table identifiers
        rms_tables = self.fields_and_tables
        table_index = {}
        fields = {}
        selected_fields = []
        selections = self.selections

        for field in rms_tables:
            table = rms_tables[field]
            identifier = None 
            if table not in table_index.values():
                identifier = 'T' + str(len(table_index)+1)
                table_index['%s' %identifier] = table

            else: 
                identifier = [i for i in table_index if table_index[i] == table][0]
            fields[field] = identifier
            if field in selections:
                selected_fields.append('%s.%s' % (identifier, field))

        self.table_index = table_index
        self.fields = fields 
        self.selected_fields = selected_fields

    def create_where_clause (self, where_list):
        # Assemble where clause
        fields = self.fields 

        where_clauses = {}
        WHERE = []
        for array in where_list:
            #print(array)
            where_clauses[len(where_clauses) +1] = array
            #break
        where_list = None
        for key, array in where_clauses.items():
            #break
            tmp = []
            shelf = []
            rms_field = array['FIELD']
            identifier = fields[ rms_field ]
            query_field = identifier + '.' + rms_field

            if array['RANGE'] == True:
                WHERE.append( "(%s BETWEEN '%s' AND '%s')" % (query_field, array['VALUES'][0], array['VALUES'][1]) )
                continue
                    
            if 'OPERATOR' in array.keys():
                for value in array['VALUES']:
                    tmp.append( '(%s %s %s)' % (query_field, array['OPERATOR'], value ) )
                shelf.append( '(%s)' % array['SPLIT'].join(tmp) )
            
            if 'WILDCARD' in array.keys():
                if array['WILDCARD'] == True: 
                    if array['EQUALS'] == True: 
                        operator = 'LIKE'
                    else: 
                        operator = 'NOT LIKE'
                    for value in array['VALUES']:
                        tmp.append( " %s %s '%s' " %(query_field, operator, value) )
                else: 
                    if array['EQUALS'] == True:
                        operator = 'IN'
                    else: 
                        operator = 'NOT IN'
                    comparison_shelf = []
                    for value in array['VALUES']:
                        comparison_shelf.append("'%s'" %value)
                    tmp.append( "%s %s (%s)" % (query_field, operator, ', '.join(comparison_shelf)) )
                    comparison_shelf = None
            shelf = ' (%s) ' % (array['SPLIT'].join(tmp))
            operator = None 
            tmp = None 
            WHERE.append(shelf)
            shelf = None
        where_clauses = None
        #print(WHERE)
        self.where = 'WHERE ( %s )' % ' AND '.join(WHERE)
        WHERE = None
    
    def create_order_by_list (self, array) :
        # Create Order list
        fields = self.fields 
        shelf = []
        for field, sort in array.items():
            identifier = fields[field]
            shelf.append('%s.%s %s' % (identifier, field, sort) )
        self.order_by = ' ORDER BY ' + ', '.join(shelf)
        shelf = None
    
    def create_query(self) :
        # Finish and create the query. Query is returned as text string
        table_index = self.table_index
        fields = self.fields 
        order_by = self.order_by
        selected_fields = self.selected_fields
        joins = self.joins
        where = self.where

        master_identifier = list(table_index)[0] #Always join to the master table 
        master = table_index[master_identifier] 
        sql = 'SELECT ' + ', '.join(selected_fields)

        for tableID in table_index: 
            #continue
            table = table_index[tableID]
            #identifier = table_index[table]

            if table == master:
                sql += ' FROM RMSPHBOBJ.%s %s ' % (table, tableID)
                continue 
            #print(table)
            
            if table in list(joins): 
                string = 'LEFT'
            else: 
                string = 'INNER'
            
            string += ' JOIN RMSPHBOBJ.%s %s ON %s.RMSFILENUM = ' % (table, tableID, master_identifier)

            if table in list(joins):
                string += "%s.%s AND %s.%s = '%s' " % (tableID, joins[table][0], tableID, joins[table][1], joins[table][2]   )
            else:
                string += "%s.RMSFILENUM " % (tableID)
            sql += string
        
        for statement in [where, order_by]:
            sql += statement
        return sql
"""
EXAMPLE: 

pq = r_query()
rms_tables = {
    'RMSACCTNUM' : 'RMSPMASTER',
    'RMSACCNTST' : 'RMSPMASTER',
    'RMSRECVRCD' : 'RMSPMASTER',
    'RMSOFFCRCD' : 'RMSPPRDBAL',
    'RMSLOANCD' : 'RMSPMASTER',
    'RMSALPHA1' : 'RMSPEXTDTL',
}
pq.set_fields_and_tables(rms_tables)

joins = {
    'RMSPLEGAL' : [ 'RMSFILENUM', 'RECORDTYPE', 'A' ],
    'RMSPEXTDTL' : [ 'RMSFILENUM', 'RECORDTYPE', 'RF' ],

}

selections = ['RMSACCTNUM', 'RMSACCNTST', 'RMSOFFCRCD']

pq.set_joins_tables(joins)
pq.set_selected_fields(selections)
pq.create_table_index_selections_and_fields()

where = [
    {
        'FIELD' : 'RMSALPHA1',
        'VALUES' : ['JUD'],
        'WILDCARD' : False,
        'RANGE' : False,
        'EQUALS' : True, 
        'SPLIT' : 'OR',
    },

    {
        'FIELD' : 'RMSACCNTST',
        'VALUES' : ['1B%', '1P%'],
        'WILDCARD' : True,
        'RANGE' : False,
        'EQUALS' : True, 
        'SPLIT' : 'OR',
    },

    {
        'FIELD' : 'RMSACCNTST',
        'VALUES' : ['DCE', 'DCF'],
        'WILDCARD' : False,
        'RANGE' : False,
        'EQUALS' : True, 
        'SPLIT' : 'AND',
    },
]

pq.create_where_clause(where)

order_by = {
    'RMSOFFCRCD' : 'DESC',
    'RMSACCNTST' : '',
}

pq.create_order_by_list(order_by)
pq.create_query()
"""

File: r_query.php
<?php 
/* 
    Class used to automate a DB2 query via lists
    Can be used with either the ODBC Functional Library or PDO

    10-12-24: 
    - Added order by Function
    - Enclosed WHERE in a SQL List for future additional criteria

    10-13-24:
    - Updated create_table_index to:
        - use the array keys stored with set_fields_and_lists
        - Skip after adding to the table index if the value is not in the select list

    10-14-24: Updated select_fields to also check the left joins table for the previous table

    [!!!!]__Major Revision__[!!!!]
    10-20-24: Re-worked create_where_clause to include negative checks, comparison operators, and wildcards

    10-22-24: Updated Table Index Identifiers to use 'T' with the count of tableIndex list + 1


*/
class rmsquery  {
    function set_fields_and_lists ( $list ) {
        // Must be passed as FIELD => TABLE ie RMSACCTNUM => RMSPMASTER
        $this->fields_and_lists = $list;

    }

    function set_selections ($list) {
        //can be a standard PHP array, list of records to select / display
        $this->fields = $list;

    }

    function create_table_index () {
        $rmsFields = $this->fields_and_lists;
        $selections = $this->fields;
        //Creates and labels tables ie RMSPMASTER A or RMSPEXTDTL B

        $tableIndex = [];
        $fields = [];
        foreach ($selections as $f) {
            $table = $rmsFields[$f];
            $identifier = NULL;
            if (array_key_exists($table, $tableIndex) === FALSE) {
                $identifier = 'T' . count($tableIndex) + 1;
                $tableIndex[$table] = $identifier;

            } else {
                $identifier = $tableIndex[$table];
            }

            if (!in_array($f, $selections)) {
                continue;
            }
        
            $fields[] = $identifier.'.'.$f;
        }
        $this->table_index = $tableIndex;
        $this->rms_fields_list = $fields;

    }

    function left_joins_table ($list) {
        $this->left_joins_table = $list;
    }

    function select_fields () {
        $joins = $this->left_joins_table ;
        $tableIndex = $this->table_index ;
        $tableSelections = [];
        foreach ($tableIndex as $table => $identifier) {
            if (array_search($table, array_keys($tableIndex)) === 0 ) {
                $tableSelections[] = sprintf("FROM RMSPHBOBJ.%s %s", $table, $identifier);
                continue;
            }
            $prevTable = array_keys($tableIndex)[0];//Always joins to RMSPMASTER
            
            $prevTableIdentifier = $tableIndex[$prevTable];
            if (array_key_exists($table, $joins) === TRUE || array_key_exists($prevTable, $joins) === TRUE ) {
                $string = 'LEFT';
            } else {
                $string = 'INNER';
            }
        
            $string .= sprintf( ' JOIN RMSPHBOBJ.%s %s ON %s.RMSFILENUM = ', $table, $identifier, $prevTableIdentifier );
        
            if (array_key_exists($table, $joins) === TRUE) {
                $joinValues = $joins[$table];
        
                $string .= sprintf("%s.%s AND %s.%s = '%s' ", $identifier, $joinValues[0], $identifier, $joinValues[1], $joinValues[2] );
            } else {
                $string .= "$identifier.RMSFILENUM ";
            }
            $tableSelections[] = $string;
        
        }

        $this->select_statement = 'SELECT ' . implode(', ', $this->rms_fields_list  ) . ' ' . implode(' ', $tableSelections);
    }

    function create_where_clause ($where_values_map) {
        /*
            Passed lists must be like:
            $where_values_map =
                [
                    [
                    'FIELD' => 'RMSOFFCRCD',
                    'VALUES' => ['CAV', 'PHR'] ,
                    'RANGE' => False,
                    'EQUALS' => True,
                    'WILDCARD' => True ,
                    'SPLIT' => 'OR',
                    ],
                    
                ] ;
        */
        $rmsFields = $this->fields_and_lists;
        $tableIndex = $this->table_index ;

        $where = [];
        foreach ($where_values_map as $key => $list) {
            $field = $list['FIELD'];
            $table = $rmsFields[$field];
            $identifier =  $tableIndex[$table];
            $shelf = [];
            $tmp = [];
        
            /* Date Range Block */
            if ($list['RANGE'] === TRUE ) {
                $shelf[] = sprintf("(%s.%s BETWEEN '%s' AND '%s')", $identifier, $field, $list['VALUES'][0], $list['VALUES'][1], );
            }
        
            /* Comparison Query Block */
            if (array_key_exists('OPERATOR', $list) === TRUE ) {
                foreach ($list['VALUES'] as $value) {
                    $tmp[] = sprintf(" (%s.%s %s %s) ", $identifier, $field, $list['OPERATOR'], $value );
                }
                $shelf[] = '( ' . implode($list['SPLIT'], $tmp) . ' )';
            }
        
            /*Wildcard and Non-Wildcard Query Block */
            if (array_key_exists('WILDCARD', $list) === TRUE ) {
                if ($list['WILDCARD'] === TRUE) {
                    if ($list['EQUALS'] === TRUE) {
                        $operator = 'LIKE';
                    } else {
                        $operator = 'NOT LIKE';
                    }
                    foreach ($list['VALUES'] as $value) {
                        $tmp[] = sprintf(" %s.%s  %s '%s' ", $identifier, $field, $operator, $value);
                    }
            
                } else {
        
                    if ($list['EQUALS'] === TRUE) {
                        $operator = 'IN';
                    } else {
                        $operator = 'NOT IN';
                    }
                    $comparison_shelf = [];
                    foreach ($list['VALUES'] as $value ) {
                        $comparison_shelf[] = "'$value'";
                    }
                    $tmp[] = sprintf("(%s.%s %s (%s)) ", $identifier, $field, $operator, implode(', ', $comparison_shelf));
                    $comparison_shelf = null;
        
                }
        
        
                $shelf[] = '( ' . implode($list['SPLIT'] ,$tmp) . ' )' ;
                $operator = NULL;
        
        
            }
        
            $tmp = NULL;
            foreach ($shelf as $s) {
                $where[] = $s;
            }
        
        
        
        }
        
        $sql =  sprintf(" WHERE ( %s ) ", implode(' AND ', $where) );
        $this->where_clauses = $sql ;

    }

    function create_order_by_list ($list) {
        $rmsFields = $this->fields_and_lists;
        $tableIndex = $this->table_index ;

        $sql = ' ORDER BY ';
        foreach ($list as $order_field => $sort_flag) {
            $sql .= sprintf( "%s.%s ", $tableIndex[  $rmsFields[$order_field] ], $order_field );
            if (ctype_alpha($sort_flag)) {
                $sql .=  $sort_flag . ' ' ;
            } 
            if ($order_field !== array_key_last($list) ) {
                $sql .= ', ';
            }
        
        }
        $this->order_by_list = $sql ;
        
    }

    function create_query () {
        $order_by = $this->order_by_list ;
        $where_clauses = $this->where_clauses;
        $select = $this->select_statement ;

        return $select .' '. $where_clauses . ' ' . $order_by;
    }


}


?>
